From e0023ee65056d45d2392ade62965d55f18dfd94b Mon Sep 17 00:00:00 2001
From: Li Lingfeng <lilingfeng3@huawei.com>
Date: Wed, 13 Jul 2022 17:52:03 +0800
Subject: [PATCH] kabi: deduplication friendly structs
Patch-mainline: Not yet, from openEuler
References: bsn#22
openEuler-commit: e0023ee65056d45d2392ade62965d55f18dfd94b
Modified-by-SEL: No


hulk inclusion
category: bugfix
bugzilla: https://gitee.com/openeuler/kernel/issues/I5DI4S
CVE: NA
Reference: https://gitlab.com/redhat/centos-stream/src/kernel/centos-stream-9/-/commit/2843fcb40c63193ad87bf3af2bac67006924bc1e

---------------------------

We don't really need the field names to be globally unique, it is enough
when they are unique in the given struct. Since structs do not generally
span mutliple files, using the line number is enough to ensure an unique
identifier. It means that we can't use two KABI_RENAME macros on the same
line but that's not happening anyway.

This allows pahole to deduplicate the type info of structs using KABI
macros, lowering the size of vmlinuz from 26M to 8.5

Signed-off-by: Li Lingfeng <lilingfeng3@huawei.com>
Reviewed-by: Zhang Yi <yi.zhang@huawei.com>
Signed-off-by: Zheng Zengkai <zhengzengkai@huawei.com>
Signed-off-by: Guoqing Jiang <guoqing.jiang@suse.com>
---
 include/linux/kabi.h | 4 +++-
 1 file changed, 3 insertions(+), 1 deletion(-)

diff --git a/include/linux/kabi.h b/include/linux/kabi.h
index a52d9fa72cfa..fe3213c0f576 100644
--- a/include/linux/kabi.h
+++ b/include/linux/kabi.h
@@ -393,6 +393,8 @@
 # define __KABI_CHECK_SIZE(_item, _size)
 #endif
 
+#define KABI_UNIQUE_ID __PASTE(kabi_hidden_, __LINE__)
+
 # define _KABI_DEPRECATE(_type, _orig)	_type kabi_reserved_##_orig
 # define _KABI_DEPRECATE_FN(_type, _orig, _args...)  \
 	_type (* kabi_reserved_##_orig)(_args)
@@ -402,7 +404,7 @@
 		_new;					  \
 		struct {				  \
 			_orig;				  \
-		} __UNIQUE_ID(kabi_hide);		  \
+		} KABI_UNIQUE_ID;			  \
 		__KABI_CHECK_SIZE_ALIGN(_orig, _new);  \
 	}
 #else
-- 
2.34.1

